src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala

```diff
@@ -0,0 +1,159 @@
+   * @param openGapProbability Penalty to start an insertion or deletion in the alignment
+   * @param closeGapProbability Penalty to end in an insertion or deletion in the alignment
+   * @return An alignment path and score
+   */
+  def align(sequence: Seq[Byte],
+            reference: Seq[Byte],
+            mismatchProbability: Double = math.exp(-4),
+            openGapProbability: Double = math.exp(-6),
+            closeGapProbability: Double = 1 - math.exp(-1)): ReadAlignment = {
+    // TODO: What are the best defaults?
+    // BWA defaults:
+    // -log mismatchProbability = 4
+    // -log openGapProbability = 6
+    // -log (1 - closeGapProbability) = 1
+
+    val sequenceLength = sequence.length
+    val referenceLength = reference.length
+    val (alignmentStates, alignmentScores) = scoreAlignmentPaths(sequence,
+      reference,
+      mismatchProbability,
+      openGapProbability,
+      closeGapProbability)
+
+    findBestPath(alignmentStates, alignmentScores, sequenceLength, referenceLength)
+  }
+
+  def scoreAlignmentPaths(sequence: Seq[Byte],
+                          reference: Seq[Byte],
+                          mismatchProbability: Double,
+                          openGapProbability: Double,
+                          closeGapProbability: Double): (DenseMatrix[AlignmentState], DenseMatrix[Double]) = {
+
+    val logMismatchPenalty = -math.log(mismatchProbability)
+    val logOpenGapPenalty = -math.log(openGapProbability)
+    val logCloseGapPenalty = -math.log(closeGapProbability)
+
+    val logContinueGapPenalty = -math.log(1 - closeGapProbability)
+    val noGapPenalty = -math.log(1 - openGapProbability)
+
+    val sequenceLength = sequence.length
+    val referenceLength = reference.length
+
+    // alignmentScores is M x N matrix
+    // M is the length of the sequence and N is the length of the reference
+    val alignmentScores = DenseMatrix.fill[Double](sequenceLength + 1, referenceLength + 1) { Double.PositiveInfinity }
+    val alignmentStates = new DenseMatrix[AlignmentState](sequenceLength + 1, referenceLength + 1)
+
+    def transitionPenalty(nextState: AlignmentState, previousState: AlignmentState, isEndState: Boolean) = {
+
+      val openGap = nextState != previousState && AlignmentState.isGapAlignment(nextState)
+      val closeGap = nextState != previousState && AlignmentState.isGapAlignment(previousState)
+      val continueGap = nextState == previousState && AlignmentState.isGapAlignment(nextState)
+      val mismatch = nextState == AlignmentState.Mismatch
+
+      (if (openGap) logOpenGapPenalty else 0) +
+        (if (closeGap) logCloseGapPenalty else 0) +
+        (if (continueGap) logContinueGapPenalty else if (mismatch) noGapPenalty + logMismatchPenalty else noGapPenalty) +
+        (if (isEndState && AlignmentState.isGapAlignment(nextState)) logCloseGapPenalty else 0)
+    }
+
+    // empty case
+    alignmentScores(0, 0) = 0
```

Comment:
What happens to the rest of the 0th row and 0th column? They get default scores of ∞, meaning we don't allow alignments that start with an insertion or deletion, right?